### ISSUE TITLE
Cannot create PoolableConnectionFactory (The connection attempt failed.)

### ISSUE THREAD

**Issue by mydeveloperplanet** [FraunhoferIOSB/FROST-Server#1122]
I am using the docker-compose file as listed in the documentation
```
version: '2'

services:
  web:
    image: fraunhoferiosb/frost-server:latest
    environment:
      - serviceRootUrl=http://localhost:8080/FROST-Server
      - http_cors_enable=true
      - http_cors_allowed.origins=*
      - persistence_db_driver=org.postgresql.Driver
      - persistence_db_url=jdbc:postgresql://database:5432/sensorthings
      - persistence_db_username=sensorthings
      - persistence_db_password=ChangeMe
      - persistence_autoUpdateDatabase=true
    ports:
      - 8080:8080
      - 1883:1883
    depends_on:
      - database

  database:
    image: postgis/postgis:11-2.5-alpine
    environment:
      - POSTGRES_DB=sensorthings
      - POSTGRES_USER=sensorthings
      - POSTGRES_PASSWORD=ChangeMe
    volumes:
      - postgis_volume:/var/lib/postgresql/data
volumes:
    postgis_volume:
```
During startup (docker-compose up) the following error occurs:
```
web_1       | 09:18:38.143 [            main] ERROR d.f.i.i.f.p.p.PostgresPersistenceManager - Could not initialise database.
web_1       | java.sql.SQLException: Cannot create PoolableConnectionFactory (The connection attempt failed.)
web_1       | 	at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:653)
web_1       | 	at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:531)
web_1       | 	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:731)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.persistence.pgjooq.utils.ConnectionUtils$ConnectionSourceBasicDataSource.getConnection(ConnectionUtils.java:206)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.persistence.pgjooq.utils.ConnectionUtils.getPoolingConnection(ConnectionUtils.java:96)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.persistence.pgjooq.utils.ConnectionUtils.getConnection(ConnectionUtils.java:77)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.persistence.pgjooq.PostgresPersistenceManager.doUpgrades(PostgresPersistenceManager.java:350)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.util.LiquibaseUtils.maybeUpdateDatabase(LiquibaseUtils.java:48)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.persistence.PersistenceManagerFactory.init(PersistenceManagerFactory.java:66)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.http.common.AbstractContextListener.contextInitialized(AbstractContextListener.java:84)
web_1       | 	at de.fraunhofer.iosb.ilt.frostserver.ContextListener.contextInitialized(ContextListener.java:37)
web_1       | 	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4768)
web_1       | 	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5230)
web_1       | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
web_1       | 	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:726)
web_1       | 	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:698)
web_1       | 	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:696)
web_1       | 	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1185)
web_1       | 	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1933)
web_1       | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
web_1       | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
web_1       | 	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
web_1       | 	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118)
web_1       | 	at org.apache.catalina.startup.HostConfig.deployDirectories(HostConfig.java:1095)
web_1       | 	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:477)
web_1       | 	at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1618)
web_1       | 	at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:319)
web_1       | 	at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
web_1       | 	at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:423)
web_1       | 	at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:366)
web_1       | 	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:946)
web_1       | 	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:835)
web_1       | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
web_1       | 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396)
web_1       | 	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386)
web_1       | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
web_1       | 	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
web_1       | 	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
web_1       | 	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919)
web_1       | 	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:263)
web_1       | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
web_1       | 	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:432)
web_1       | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
web_1       | 	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:927)
web_1       | 	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
web_1       | 	at org.apache.catalina.startup.Catalina.start(Catalina.java:772)
web_1       | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
web_1       | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
web_1       | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
web_1       | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
web_1       | 	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:345)
web_1       | 	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:476)
web_1       | Caused by: org.postgresql.util.PSQLException: The connection attempt failed.
web_1       | 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:331)
web_1       | 	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
web_1       | 	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
web_1       | 	at org.postgresql.Driver.makeConnection(Driver.java:400)
web_1       | 	at org.postgresql.Driver.connect(Driver.java:259)
web_1       | 	at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:52)
web_1       | 	at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:374)
web_1       | 	at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:106)
web_1       | 	at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:649)
web_1       | 	... 51 common frames omitted
web_1       | Caused by: java.net.SocketTimeoutException: connect timed out
web_1       | 	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
web_1       | 	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:412)
web_1       | 	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:255)
web_1       | 	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:237)
web_1       | 	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
web_1       | 	at java.base/java.net.Socket.connect(Socket.java:609)
web_1       | 	at org.postgresql.core.PGStream.createSocket(PGStream.java:241)
web_1       | 	at org.postgresql.core.PGStream.<init>(PGStream.java:98)
web_1       | 	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:109)
web_1       | 	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
web_1       | 	... 59 common frames omitted
web_1       | 09:18:38.145 [            main] INFO  d.f.i.i.f.p.PersistenceManagerFactory - Database-update not successful, trying again later.
```
I have tried to connect with psql into the database (using the docker container IP) and I can connect to the database with user sensorthings and password ChangeMe.

Using the url http://localhost:8080/FROST-Server/DatabaseStatus and updating the database returns an identical error. 

There is, however, also a SEVERE log statement during startup, but do not know whether this has any influence on this problem.
```
web_1       | 09-Sep-2022 09:18:24.120 SEVERE [main] org.apache.catalina.startup.HostConfig.beforeStart Unable to create directory for deployment: [/usr/local/tomcat/conf/Catalina/localhost]
```

What is going wrong here?

**Comment by mydeveloperplanet** [comment]
I have some additional information:
It works for a colleague of mine who has a Windows machine and uses Docker desktop.

I work with Ubuntu with probably a few non-standard security improvements:
* My user does not have the default userid 1000
* My umask is set in order that files are not by default readable by 'other'

I have encountered in the past problems with docker containers because they assume that userid 1000 is used on the host system and that the file permissions are set in order that files are readable by 'other'. In all cases, this had to be solved in the Docker image itself.

**Comment by hylkevds** [comment]
There seems to be a networking problem in your Docker setup. The web container can't reach the database container.
Does Docker create the frost_default network? (`docker network ls`)

The `Unable to create directory for deployment` error should not be a problem. It just means Tomcat can't copy the context.xml file and will just use the original from the jar.
Running as a different user also works fine.

**Comment by mydeveloperplanet** [comment]
The network seems to be ok
```bash
$ docker network ls
NETWORK ID     NAME            DRIVER    SCOPE
fd5324bc60b4   bridge          bridge    local
c75deada4b9d   frost_default   bridge    local
17c1d3b47c87   host            host      local
f9bcbb8a4c43   none            null      local
```

**Comment by mydeveloperplanet** [comment]
I have a working setup now, I only do not know how I did it ;-)
I got a tip that it might be caused by the default IP range docker is using, see https://serverfault.com/questions/916941/configuring-docker-to-not-use-the-172-17-0-0-range

Changing the IP range did not work for some reason and this was due to the fact that I had installed docker via snap (I am working with Ubuntu). Configuring docker installed via snap gave some other issues, so I decided to remove the docker installation and install it according to the installation instructions provided by docker.
And now it works. But, I am also working now at my home location and maybe (but I do not remember this) it did not work before at my work location. But then, the info in the link above should work.
